bag_verifier/src/main/scala/uk/ac/wellcome/platform/archive/bagverifier/fixity/azure/AzureDynamoTags.scala

```diff
@@ -0,0 +1,84 @@
+package uk.ac.wellcome.platform.archive.bagverifier.fixity.azure
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
+import com.azure.storage.blob.BlobServiceClient
+import org.scanamo.auto._
+import org.scanamo.syntax._
+import org.scanamo.{Scanamo, Table}
+import uk.ac.wellcome.storage.azure.AzureBlobLocation
+import uk.ac.wellcome.storage.dynamo.DynamoConfig
+import uk.ac.wellcome.storage.tags.Tags
+import uk.ac.wellcome.storage._
+
+// Although Azure Blob Storage supports S3-style tags, they're only in preview
+// and not available in our region yet.
+// https://docs.microsoft.com/en-us/azure/storage/blobs/storage-manage-find-blobs?tabs=azure-portal#regional-availability-and-storage-account-support
+//
+// Using DynamoDB to store tags about blobs in Azure is a stopgap until we get
+// access to first-class Azure tags.
+class AzureDynamoTags(dynamoConfig: DynamoConfig)(
+  implicit
+  blobServiceClient: BlobServiceClient,
+  dynamoClient: AmazonDynamoDB
+) extends Tags[AzureBlobLocation] {
+  case class DynamoTagsEntry(id: String, tags: Map[String, String])
+
+  private val table: Table[DynamoTagsEntry] =
+    Table[DynamoTagsEntry](dynamoConfig.tableName)
+
+  private val scanamo = Scanamo(dynamoClient)
+
+  private def exists(location: AzureBlobLocation): Boolean =
+    blobServiceClient
+      .getBlobContainerClient(location.container)
+      .getBlobClient(location.name)
+      .exists()
+
+  override protected def writeTags(
+    location: AzureBlobLocation,
+    tags: Map[String, String]
+  ): Either[WriteError, Map[String, String]] =
+    if (exists(location)) {
+      val ops = if (tags.isEmpty) {
+        table.delete('id -> location.toString())
+      } else {
+        table.put(DynamoTagsEntry(id = location.toString(), tags = tags))
+      }
+
+      scanamo.exec(ops) match {
```

Comment:
I know it's a draft, but why do we use scanamo direclty instead of the store interface?

Reply:
The Dynamo stores we've already got assume an ID and a version (even if the thing isn’t versioned); we don’t enforce versioning constraints on tags in S3.

Also I’m not entirely convinced those stores would handle tags correctly – in particular, if you write `CacheEntry(id = "1", tags = Map.empty)`, what actually gets written to DynamoDB is `"id" -> "1", "tags" -> null`, and then Scanamo has issues serialising it back out again.